Understand changes of running-time for each task

ystopia opened the pull request:
I added some codes for visualization of how long each task takes running-time until it reaches the next status (failed or done).
1. At "{base_url}/tasklist", all completed(failed or done) tasks are shown.
2. At "{base_url}/tasklist", a user can select one specific task to see how its running-time has changed over time.
3. At "{base_url}/tasklist/{task_name}", it visualizes a multi-bar graph that represents the changes of the running-time for a selected task up to the next status (failed or done).
This visualization let us know how the running-time of the specific task has changed over time.

gpoulin commented on luigi/server.py:
```diff
@@ -14,10 +14,19 @@
-"""
-Simple REST server that takes commands in a JSON payload.
-Interface to the :py:class:`~luigi.scheduler.CentralPlannerScheduler` class.
-"""
```
Why removing the module doctring?

ystopia replied:
Interesting. I didn't touch anything about those lines but something happened. I will change them to the original status!

erikbern replied:
You probably need to rebase on latest master

gpoulin commented on luigi/templates/history.html:
```diff
@@ -0,0 +1,117 @@
+		var running_data = data.RUNNING;
+		var done_data = data.DONE;
+		var failed_data = data.FAILED;
+		var all_results = unescape("{{ tableResults }}");
```
Could it be an ajax call? I personally don't trust javascript templating.

Tarrasch replied:
Yes. And sorry, no test ==> no merge I think. :)

Looking forward to an iteration on this PR :)

ystopia replied:
Oh, I might need to use Travis CI test for this, right?

gpoulin commented on luigi/templates/layout.html:
```diff
@@ -0,0 +1,111 @@
+	  }
+      .lead { margin-top: -17px; margin-bottom: 13px; }
+
+    </style>
```
This could be in a css file? It's easier to override after.

gpoulin commented:
Look interesting. I think having ETA is often requested feature[citation needed]. Could you try to make Travis happy?

https://travis-ci.org/spotify/luigi/jobs/51938806#L316
https://travis-ci.org/spotify/luigi/jobs/51938804#L119
https://travis-ci.org/spotify/luigi/jobs/51938807#L138

Tarrasch commented:
You'll need to clean up the git history here. However, we won't be able to merge this until next Thursday (I'll get back about this later).

erikbern commented on luigi/server.py:
```diff
@@ -73,6 +92,62 @@ def get_template_path(self):
+            task_seq = task['id']
+            task_name = task['name']
+            tasks[task_seq] = str(task_name)
+        for task in task_history:
```
Seems unnecessary to query all tasks here, maybe I'm missing something but can't you do `select id, task_id, event_name, ts from task_events where task_id = %s`?

ystopia replied:
Oh, this was because I thought it's useful to see the name of the task in addition to its task id. (e.g. The task named "Loader" can have event named "Done", "Failed", and "Running.")

erikbern commented on luigi/server.py:
```diff
@@ -73,6 +92,62 @@ def get_template_path(self):
+def from_utc(utcTime, fmt="%Y-%m-%d %H:%M:%S.%f"):
+    """convert UTC time string to time.struct_time: change datetime.datetime to time, return time.struct_time type"""
+    time_struct = datetime.datetime.strptime(utcTime, fmt)
+    date = int(time.mktime(time_struct.timetuple())) * 1000
```
Minor point but the millisecond standard is a Java/Javascript convention that doesn't exist in other language. I would prefer if the millisecond conversion takes place in the frontend instead.

ystopia replied:
Oh I see. Thanks. I will fix it by inserting conversion to millisecond in the Javascript side, and returning unix timestamp in seconds in this python code!

erikbern commented:
This is great – one step towards ETA and stuff.

Is this enabled by default (if the task history db is enabled)?

Is it backwards compatible?

ystopia commented:
Yes I enabled this with 3 html files by default if the task history db is enabled.
I'm sorry but what do you mean by "backwards compatible"?

gpoulin2 commented on luigi/server.py:
```diff
@@ -73,6 +92,62 @@ def get_template_path(self):
+        chartResults = {}
+        tableResults = []
+        # get task history db in the same path
+        engine = create_engine('sqlite:///.task_history.db', echo=True)
```
Don't assume sqlite storage. https://github.com/spotify/luigi/blob/master/luigi/db_task_history.py#L63

ystopia replied:
Thanks. I will replace the sqlite connection string part with this.

erikbern replied:
just wanted to point out this is still remaining

ystopia replied:
I came back to this work and commented to the html files. But after the version upgrading of luigi, I faced with the db connection issue. I'm trying to follow the current way that luigi connects to db with sqlalchemy.

gpoulin2 commented on luigi/server.py:
```diff
@@ -73,6 +92,62 @@ def get_template_path(self):
+        engine = create_engine('sqlite:///.task_history.db', echo=True)
+        conn_sqlite = engine.connect()
+        task_query = "select id, name from tasks"
+        history_query = "select id, task_id, event_name, ts from task_events"
```
Maybe use the sqlalchemy ORM instead. Or simply use `find_all_by_name` and lazy load the events. http://docs.sqlalchemy.org/en/rel_0_9/orm/tutorial.html

erikbern commented:
With backwards compatibility I meant does it support rendering (old) tasks for which these attributes aren't set?

ystopia commented:
Yes it does. It supports all old and current tasks stored in a task history db to be rendered since they have their status with "Failed", "Running", or "Done".

erikbern commented:
Awesome. I retriggered the unit tests since there was some intermittent issue. Great if you can address the remaining issues.

erikbern commented:
Great that tests are successful now

Tarrasch commented:
Awesome, can you go through one iteration of the code code review @ystopia? :)

erikbern commented:
@ystopia I'm happy to merge this is if you address some of the comments in the commit. There's also a merge conflict so you need to rebase on latest master.

ystopia commented:
Oh, sorry for being late. Busy in doing other stuffs...I will put some comments in the source code and in the commit. Rebasing on the latest master too. :)

erikbern commented:
Cool no stress

On Tuesday, March 10, 2015, ystopia <notifications@github.com> wrote:

> Oh, sorry for being late. Busy in doing other stuffs...I will put some
> comments in the source code and in the commit. Rebasing on the latest
> master too. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/luigi/pull/783#issuecomment-78028538>.
>

erikbern commented:
Ping @ystopia – would be great if you can update the PR when you have a chance!

erikbern commented:
looks good but please rebase on latest master and squash some of your commits

maybe https://cdnjs.com/libraries/d3/ ?

ystopia commented:
Yes.
https://cdnjs.com/libraries/d3/
https://cdnjs.com/libraries/nvd3

erikbern commented:
I re-triggered the build, should pass now

erikbern commented:
Something is a bit weird with the second commit though ("Merge remote-tracking...")

Did you rebase onto master?

ystopia commented:
I rebased onto ystopia:master, then synced my github fork (ystopia:master) with spofity:master.
Everyting is up-to-date.

erikbern commented:
Nice! I'm not sure why that second commit showed up. I merged and the log looks good.

Thanks!

ystopia commented:
Thank you!

ystopia commented:
Hi,
I would like to pull request on following changes, which I just have pushed to ystopia:master.
I don't know why it does not automatically pull request to spotify:master.
Several changes are,
1. added copyright & author & description info to the db_task_history.py
2. re-added three template files (header.html, recent.html, show.html should stay there but I removed them by accident a few days ago. Sorry about that.)
3. modified the bootstrap lib path in layout.html

erikbern commented:
you can try to push to another branch (eg ystopia:foo)
